mango-core/src/main/scala/org/bdgenomics/mango/io/BedReader.scala

```diff
@@ -0,0 +1,175 @@
+import org.apache.spark.SparkContext
+import org.bdgenomics.adam.models.ReferenceRegion
+import org.bdgenomics.mango.models.LazyMaterialization
+import org.bdgenomics.utils.misc.Logging
+import htsjdk.tribble.index.{ Index, IndexFactory }
+
+import scala.collection.JavaConversions._
+import org.bdgenomics.formats.avro.Feature
+import org.bdgenomics.adam.rdd.ADAMContext._
+import org.bdgenomics.adam.rdd.feature.FeatureDataset
+
+object BedReader extends GenomicReader[FeatureCodecHeader, Feature, FeatureDataset] with Logging {
+
+  def suffixes = Array(".bed", ".bed.gz", ".narrowPeak")
+
+  def load(fp: String, regionsOpt: Option[Iterable[ReferenceRegion]], local: Boolean): Tuple2[FeatureCodecHeader, Array[Feature]] = {
+
+    val codec: BEDCodec = new BEDCodec()
+
+    // if not valid throw Exception
+    if (!isValidSuffix(fp)) {
+      invalidFileException(fp)
+    }
+
+    // reader must be called before createIndex is called
+    val reader = AbstractFeatureReader.getFeatureReader(fp, codec, false)
+
+    if (local) {
+      // create index file, if it does not exist
+      createIndex(fp, codec)
+    }
+
+    val header = reader.getHeader().asInstanceOf[FeatureCodecHeader]
+    val dictionary = reader.getSequenceNames()
+
+    val results =
+      if (regionsOpt.isDefined) {
+        val queries =
+          if (!dictionary.isEmpty) {
+
+            // modify chr prefix, if this file uses chr prefixes.
+            val hasChrPrefix = dictionary.get(0).startsWith("chr")
+
+            regionsOpt.get.map(r => {
+              LazyMaterialization.modifyChrPrefix(r, hasChrPrefix)
+            })
+          } else {
+            // take all possible combinations of ReferenceRegions
+            regionsOpt.get.map(r => {
+              Iterable(LazyMaterialization.modifyChrPrefix(r, true), LazyMaterialization.modifyChrPrefix(r, false))
+            }).flatten
+          }
+
+        if (reader.isQueryable) {
+          val tmp = queries.map(r => reader.query(r.referenceName, r.start.toInt, r.end.toInt).toList)
+          reader.close()
+          tmp.flatten
+        } else {
+          // in the case where file is remote and doesn't have an index, still want to get data
+          val iter: Iterator[BEDFeature] = reader.iterator()
+          reader.close()
+          iter.map(r => {
+            val overlaps = queries.filter(q => q.overlaps(ReferenceRegion(r.getContig, r.getStart, r.getEnd))).size > 0
+            if (overlaps)
+              Some(r)
+            else None
+          }).flatten
+        }
+      } else {
+        // no regions specified, so return all records
+        val iter: Iterator[BEDFeature] = reader.iterator()
+        reader.close()
+        iter
+      }
+
+    // map results to ADAM features
+    val features = results.map(r => Feature.newBuilder()
+      .setFeatureType(r.getType)
+      .setReferenceName(r.getContig)
+      .setStart(r.getStart.toLong - 1) // move to 0 indexed to match HDFS results
```

Comment:
BED format uses 0-based, closed-open intervals.  This shouldn't be required, unless htsjdk converts them to 1-based, fully closed intervals.

https://github.com/heuermh/bdg-formats/blob/docs/docs/source/features.md

Reply:
This is required to match the ADAM formats. One of them is wrong, so I matched the ADAM formats